utils/clflags.ml

```diff
@@ -397,53 +399,123 @@ module Extension = struct
     | "polymorphic_parameters" -> Some Polymorphic_parameters
     | "immutable_arrays_experimental" -> Some Immutable_arrays
     | _ -> None
 
-  let disable_all_extensions = ref false             (* -disable-all-extensions *)
+  let of_string_exn extn =
+    match of_string extn with
+    | Some extn -> extn
+    | None -> raise (Arg.Bad(Printf.sprintf "Extension %s is not known" extn))
 
-  let disable_all () =
-    disable_all_extensions := true;
-    match !extensions with
-    | [] -> ()
-    | ls ->
-      raise (Arg.Bad(Printf.sprintf
-        "Compiler flag -disable-all-extensions is incompatible with \
-         the enabled extensions: %s"
-        (String.concat "," (List.map to_string ls))))
+  (* We'll do this in a more principled way later *)
+  let is_erasable = function
+    | Local ->
+        true
+    | Comprehensions
+    | Include_functor
+    | Polymorphic_parameters
+    | Immutable_arrays ->
+        false
+
+  module Universe = struct
+    (** Which extensions can be enabled? *)
+    type t =
+      | No_extensions
+      | Only_erasable
+      | Any
+
+    let compiler_options = function
+      | No_extensions -> "flag -disable-all-extensions"
+      | Only_erasable -> "flag -only-erasable-extensions"
+      | Any           -> "default options"
+
+    let is_allowed t ext = match t with
+      | No_extensions -> false
+      | Only_erasable -> is_erasable ext
+      | Any           -> true
+  end
+
+  (* Mutable state.  Invariants:
 
-  let enable extn =
-    if !disable_all_extensions then
+     (1) [!extensions] contains at most one copy of each extension.
+
+     (2) Every member of [!extensions] satisfies [Universe.is_allowed !universe].
+         (For instance, [!universe = No_extensions] implies
+         [!extensions = []]). *)
+  let extensions = ref default_extensions (* -extension *)
+  let universe   = ref Universe.Any       (* -only-erasable-extensions,
+                                             -disable-all-extensions *)
+
+  let set extn ~enabled =
+    if not (Universe.is_allowed !universe extn) then
       raise (Arg.Bad(Printf.sprintf
-        "Cannot enable extension %s: \
-         incompatible with compiler flag -disable-all-extensions"
-        extn));
-    match of_string (String.lowercase_ascii extn) with
-    | Some extension ->
-        if not (List.exists (equal extension) !extensions) then
-          extensions := extension :: !extensions
-    | None ->
-        raise (Arg.Bad (Printf.sprintf "Unknown extension \"%s\"" extn))
-
-  let is_enabled ext =
-    not !disable_all_extensions
-    && (List.mem ext default_extensions
-        || List.mem ext !extensions)
+        "Cannot %s extension %s: incompatible with %s"
+        (if enabled then "enable" else "disable")
+        (to_string extn)
+        (Universe.compiler_options !universe)));
+    if enabled then begin
+      if not (List.exists (equal extn) !extensions) then
+        extensions := extn :: !extensions
```

Comment:
Should this just be a set instead of a list? Maybe not, because it's so small, so constant factors overwhelm asymptotic ones. In this case, though, just use `filter` below, without the early cutoff. That is, either this is small and simple, or it's big and needs to be efficient. Not sure which is better, but I think small and optimized is a poor choice.

Reply:
I think you are probably correct, so I've made the change, but it galls me slightly because it's so obviously inefficient.  (By, like, 2 things.  In a case which will ~never be called.)  Leaving this comment to get that out of my system and making the code change.